src/Functional.CQS.AOP.IoC.PureDI.MetricsCapturing.Tests/ResultCommandHandlerMetricsCapturingDecoratorTests.cs

```diff
@@ -0,0 +1,102 @@
+using System;
+using AutoFixture;
+using AutoFixture.Xunit2;
+using FakeItEasy;
+using Functional.CQS.AOP.CommonTestInfrastructure.DummyObjects;
+using Functional.CQS.AOP.MetricsCapturing;
+using Xunit;
+
+namespace Functional.CQS.AOP.IoC.PureDI.MetricsCapturing.Tests
+{
+	public class CommandHandlerMetricsCapturingDecoratorTests
+	{
+		[Theory]
+		[CommandHandlerCompletesSuccessfully]
+		public void ShouldCaptureResultAndElapsedTime(
+			CommandHandlerMetricsCapturingDecorator<DummyCommandThatSucceeds, DummyCommandError> sut,
+			IMetricsCapturingStrategyForCommand<DummyCommandThatSucceeds, DummyCommandError> metricsCapturingStrategy)
+		{
+			var command = new DummyCommandThatSucceeds();
+			sut.Handle(command);
+
+			A.CallTo(() => metricsCapturingStrategy.OnInvocationStart(command)).MustHaveHappenedOnceExactly();
+			A.CallTo(() => metricsCapturingStrategy.OnInvocationCompletedSuccessfully(command, A<Result<Unit, DummyCommandError>>._, A<TimeSpan>._)).MustHaveHappenedOnceExactly();
+			A.CallTo(() => metricsCapturingStrategy.OnInvocationException(command, A<Exception>._, A<TimeSpan>._)).MustNotHaveHappened();
+		}
+
+		[Theory]
+		[CommandHandlerThrowsException]
+		public void ShouldCaptureExceptionAndElapsedTime(
+			CommandHandlerMetricsCapturingDecorator<DummyCommandThatSucceeds, DummyCommandError> sut,
+			IMetricsCapturingStrategyForCommand<DummyCommandThatSucceeds, DummyCommandError> metricsCapturingStrategy)
+		{
+			var command = new DummyCommandThatSucceeds();
+			Result.Try(() => sut.Handle(command));
+
+			A.CallTo(() => metricsCapturingStrategy.OnInvocationStart(command)).MustHaveHappenedOnceExactly();
+			A.CallTo(() => metricsCapturingStrategy.OnInvocationCompletedSuccessfully(command, A<Result<Unit, DummyCommandError>>._, A<TimeSpan>._)).MustNotHaveHappened();
+			A.CallTo(() => metricsCapturingStrategy.OnInvocationException(command, A<Exception>._, A<TimeSpan>._)).MustHaveHappenedOnceExactly();
+		}
+
+		#region Arrangements
+
+		private abstract class CommandHandlerMetricsCapturingDecoratorTestsArrangementBase : AutoDataAttribute
+		{
+			protected CommandHandlerMetricsCapturingDecoratorTestsArrangementBase(Func<Result<Unit, DummyCommandError>> resultFactory)
+				: base(() => new Fixture()
+					.Customize(new CommandHandlerCustomization(resultFactory))
+					.Customize(new MetricsCapturingStrategyCustomization()))
+			{
+
+			}
+		}
+
+		private class CommandHandlerCompletesSuccessfully : CommandHandlerMetricsCapturingDecoratorTestsArrangementBase
+		{
+			public CommandHandlerCompletesSuccessfully()
+				: base(() => Result.Success<Unit, DummyCommandError>(Unit.Value))
```

Comment:
Other tests may be doing this as well, but this is the first one where I noticed it.
It looks like when your handler returns `Result<TSuccess, TFailure>`, you test the `Result.Succeed` and `Exception-thrown` cases, but _not_ the `Result.Fail` case.

It may be worth testing the `Result.Fail` case for completeness.

Reply:
`Result.Fail` is covered by the `Result.Succeed` case (the value of the `Result` doesn't matter...  only that a `Result` was received).  Successful execution != `Result.Success`, only that an exception was not thrown.